Optimise autocomplete query

romeovs opened the pull request:
Partially resolves https://github.com/metabase/metabase/issues/30588

### Description

This PR implements some of the suggestions made in https://github.com/metabase/metabase/issues/30588#issuecomment-1536676195, as far as `toucan2` will take them.

This is what's implemented:

1. Optimizing the auto-suggestion queries (at least 10x faster is needed): 
    - [x] `A materialized view might work very well here, to be tested`, or an index on a `lower(metabase_field.name)` (only for pgsql though)
    - [x] `Checking if the table is active in the WHERE`,
    - [x] `Duplicating the .active checks in the JOIN ON … condition`.
    - [x] `The LEFT JOIN is not necessary (and should be changed to a INNER JOIN)`
2. `Lower the spam of queries with some trick`. This is done in #37852

### How to verify

It's hard to verify the extreme memory usage locally, but we can at least verify these changes did not break the autocomplete endpoint:

1. `SQL Query` → `Sample Dataset`
2. Start typing a query like `SELECT SEA`
3. The autocompletion dropdown should return `SEATS`

### Checklist

I've added no extra tests for this issue.

romeovs committed:
Filter for active tables only

romeovs committed:
Switch to more efficient inner-join

piranha commented:
Would be nice to have some query which exhibits performance problems, right now looking at `explain analyze` is almost useless since it's so fast for everything I'm trying... Have you seen slowness like this on stats perhaps?

piranha commented:
As for point number three, you can write `:inner-join [[:metabase_table :table] [:and [:= :table.id :metabase_field.table_id] :table.active]]` and this will work.

sansolovyov committed:
move check for active table inside the join

sansolovyov committed:
add index for autocomplete search on fields

sansolovyov committed:
h2 can't do indexes on functions

sansolovyov committed:
ok, checking for active fields breaks tests

and I can't find the reason - but it has no major impacts on performance after adding an index, at least on stats

sansolovyov committed:
no indexes on expression for me, at least I can set Cache-Control

piranha commented:
@metabase/core-backend-components this is ready for review

qnkhuat commented:
Could you add note on why the cache control change is needed?

sansolovyov committed:
explain caching setup

piranha commented:
@qnkhuat done, how'd you like it?

crisptrutski commented on src/metabase/api/database.clj:
```diff
@@ -620,11 +621,16 @@
-      (autocomplete-suggestions id (str "%" substring "%"))
-      prefix
-      (autocomplete-suggestions id (str prefix "%")))
+    {:status  200
+     ;; Presumably user will repeat same prefixes many times writing the query,
+     ;; so let them cache response to make autocomplete feel fast. 60 seconds
+     ;; is not enough to be a nuisance when schema or permissions change. Cache
+     ;; is user-specific since we're checking for permissions.
+     :headers {"Cache-Control" "public, max-age=60"
+               "Vary"          "Cookie"}
+     :body    (cond
+                substring (autocomplete-suggestions id (str "%" substring "%"))
+                prefix    (autocomplete-suggestions id (str prefix "%")))}
```
nit: I'm a sucker for small methods and avoiding mixing too many levels of abstraction. I'd break out a small `with-user-specific-cache-header` method which takes the body.

piranha replied:
IDK, I'd rather this be "in the face" until a general pattern emerges...

crisptrutski replied:
Rule of 3? Fine 😄

crisptrutski commented on src/metabase/api/database.clj:
```diff
@@ -551,10 +551,11 @@
-              :limit     limit}))
+             {:order-by   [[[:lower :metabase_field.name] :asc]
+                           [[:lower :table.name] :asc]]
+              :inner-join [[:metabase_table :table] [:and :table.active
```
It might be worth commenting that it was (counter-intuitively) faster to have the "active" check within the join, in case someone comes to optimize again.

piranha replied:
Hah, it felt intuitive for me - you get less data to join this way 😁 But sure, I can leave a comment there.

crisptrutski replied:
Probably you're right, my intuition is quite rusty and I didn't give much thought to it 😆

crisptrutski commented on src/metabase/api/database.clj:
```diff
@@ -551,10 +551,11 @@
              :%lower.metabase_field/name     [:like (u/lower-case-en search-string)]
```
It might be worth commenting that an index on `lower(name)` dramatically speeds the query up in DBs that support it, but we left it for now since Maria and H2 do not support indexes on computed fields. Perhaps they'll catch up in future? I also wonder if there are other use cases for a lowercase username, we could always materialize this as another column.

crisptrutski commented on test/metabase/api/database_test.clj:
```diff
@@ -683,6 +685,13 @@
                                         ["CATEGORY_ID" "VENUES :type/Integer :type/FK"]]}]
         (is (= expected (prefix-fn (mt/id) prefix))))
+      (testing " returns sane Cache-Control headers"
+        (is (=? {"Cache-Control" "public, max-age=30"
```
This needs to be changed to 60 now

sansolovyov committed:
add back index at least for postgres

sansolovyov committed:
add comment about index in pgsql
